sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -90,18 +92,27 @@ class QueryExecution(val sparkSession: SparkSession, val logical: LogicalPlan) {
    * Prepares a planned [[SparkPlan]] for execution by inserting shuffle operations and internal
    * row format conversions as needed.
    */
-  protected def prepareForExecution(plan: SparkPlan): SparkPlan = {
-    preparations.foldLeft(plan) { case (sp, rule) => rule.apply(sp) }
-  }
+  protected def prepareForExecution(plan: SparkPlan): SparkPlan = execute(plan)
+
+  private val fixedPoint = FixedPoint(sparkSession.sessionState.conf.optimizerMaxIterations)
 
   /** A sequence of rules that will be applied in order to the physical plan before execution. */
-  protected def preparations: Seq[Rule[SparkPlan]] = Seq(
-    python.ExtractPythonUDFs,
-    PlanSubqueries(sparkSession),
-    EnsureRequirements(sparkSession.sessionState.conf),
-    CollapseCodegenStages(sparkSession.sessionState.conf),
-    ReuseExchange(sparkSession.sessionState.conf),
-    ReuseSubquery(sparkSession.sessionState.conf))
+  override protected def batches: Seq[Batch] = Seq(
```

Comment:
How about we add a `PhysicalOptimizer` to do these things? then we can simply write `lazy val executedPlan: SparkPlan = physicalOptimizer.execute(sparkPlan))` instead of `lazy val executedPlan: SparkPlan = prepareForExecution(sparkPlan)`